### DISCUSSION TITLE
Meta Web Framework in WASM/WASI using WasmEdge

### DISCUSSION THREAD

**Discussion by proohit** [WasmEdge/WasmEdge#2233] [General]
Hi WasmEdge community, 

Thanks for the great works and progress in WASM/WASI you are doing.

For my master's thesis, I want to write a web framework that is usable from several programming languages, but has one core written in Rust and compiled to WASM. As WasmEdge supports sockets and network functions, I wanted to use it as the underlying runtime.

Now, when searching through the docs and code, I could not understand if the host SDKs for Go / Node etc. actually support those network functions. I understand that the runtime itself does implement them and can thus be used as a CLI app. But do the SDKs also provide bindings to use them or is that even necessary? Or does the WASM binary just get passed to the runtime and all imports happen there?

Also to ensure easier integration, do users of said framework have to install WasmEdge **manually**? In the node SDK, the runtime can be installed with npm only, but does that apply to e.g. Go as well?

Thanks in advance!

**Comment by juntao** [comment]
You can compile an HTTP server or client into Wasm and have it run inside WasmEdge. The host app just need to call the Wasm function. There is no need for the host app to provide networking. 

You can embed a distribute the WasmEdge runtime library and SDK in your host app. The user does not need to download and install WasmEdge manually.

**Comment by proohit** [comment]
@juntao and team

I've stumbled upon a new challenge here and would be glad to gain some new insights.

I'm trying to call **dynamically added** functions from inside a WebAssembly module, which are passed from the host environment to the module. Think about it like an _add callback_ mechanism, or rpc. There can be many functions added, which are not known at compile time. In regards of this discussion, I'm trying to enable registering request handlers, which can be registered with a path and get passed a request and are expected return a response, both in JSON. 

Initially, I thought about "plain" function imports but they would have to be declared beforehand in the wasm module. As the functions can be dynamically added, this falls out of order. Then I crossed over reference types which are implemented in WasmEdge, or the successor proposal for typed function references and wondered if that was the way to go. Is there a way to access such a function from the wasm module, i.e. from Rust?

Another way I'm thinking of is to reference the function somewhere in wasm module memory and use function pointers to call it from inside the wasm module. This sounds as hacky as it propably is, so I wanted to ask for a better direction here.

Thanks in advance!

Here's a little practical example of what I'm trying to achieve:

```
## host environment

# initiate wasmedge
vm = ...

#implement some function
some_fn()

# register callback function or pass function to module
vm.add_function("some_fn", some_fn)
```

```
## wasm module (in Rust)

# get function somehow, e.g. via name
some_fn = ...

# call imported function
some_fn()
```

**Reply by juntao** [reply]
Can you provide some pseudo code examples to illustrate what you want to accomplish? Thanks.

**Reply by proohit** [reply]
Yes sure. 

Let's say there is a function signature `(req: Request) -> Response` that gets a request and should return a response. This type of function is being called when its' request path matched (see mapping in guest implementation). There may be multiple different implementations of that signature (see host functions `handle_get_some_path` and `handle_get_some_other_path`), depending on the path. The implementations are dynamically provided from the host (see `register_handler` in host) and called from inside the WebAssembly module (see `mappedFn()` in module).

```
# host

fn handle_get_some_path(req: Request) Response { ... }

fn handle_get_some_other_path(req: Request) Response { ... }

fn startModule() {
   ## initiate wasmedge
   vm  = ...

   mapping = [["/path", handle_get_some_path], ["/other-path", handle_get_some_other_path]]

   ## register functions for later use. don't mind exact types for now
   vm.Execute("register_handler", mapping)

   vm.Execute("start")
}
```

```
# wasm / guest

fn register_handler(mapping) {
   ## save path-handler mapping from host somewhere to map later
   globalPathMapping = mapping
}

fn start() {
   ## some http server code

   path = request.path

   ## get the handler function for that path that has been added/registered from host
   mappedFn = globalPathMapping.find(path)

   ## and execute it so that the implementations from host get executed
   response = mappedFn(request)
} 
```

I think the one thing that makes WebAssembly imports useless here is that the function type can be implemented multiple times, so I cannot declare inside the module during compile time.
